### ISSUE TITLE
systemd-resolved "now complete with <network-down> from none" while network is up

### ISSUE THREAD

**Issue by wdoekes** [systemd/systemd#15532]
**systemd version the issue has been seen with**
> 237-3ubuntu10.39

**Used distribution**
 > Ubuntu Bionic

**Expected behaviour you didn't see**
 > resolving of hostnames through 127.0.0.53

**Unexpected behaviour you saw**
 > SERVFAIL returned by systemd-resolved without attempting any lookups

I suspect this is because systemd-resolved for some reason thinks there is no network. I'm sorry if the cause is elsewhere, but I found _nothing_ on the internet that resembles this issue.

**Steps to reproduce the problem**

- Set up Routing on the Host (RotH, e.g. https://codingpackets.com/blog/linux-routing-on-the-host-with-frr/ )
- Observe the local IP on the `lo` device:
```
# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet 10.x.x.39/32 brd 10.x.x.39 scope global lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
...
4: enp67s0f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
...
```
- Observe the routes in vtysh:
```
# vtysh -c 'show ip route' | tail -n3
B>* 0.0.0.0/0 [20/0] via fe80::xxxx:xxxx:xxxx:f5e0, enp67s0f0, 1d17h49m
  *                  via fe80::xxxx:xxxx:xxxx:c1f, enp67s0f1, 1d17h49m
C>* 10.x.x.39/32 is directly connected, lo, 1d17h52m
```
- Observe that ping works:
```
# ping -c1 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=51 time=4.00 ms
```
- Manual DNS so it shows up in the global list (systemd-resolved refuses to get them from the `lo` device):
```
# grep ^DNS /etc/systemd/resolved.conf 
DNS=8.8.8.8

# systemd-resolve --status
Global
         DNS Servers: 8.8.8.8
...
```
- Do a lookup (with tcpdump in the background):
```
# dig -t a github.com @127.0.0.53 +short
15:13:45.882783 IP 127.0.0.1.58329 > 127.0.0.53.53: 46355+ [1au] A? github.com. (51)
15:13:45.883121 IP 127.0.0.53.53 > 127.0.0.1.58329: 46355 ServFail 0/0/1 (39)
```
- Bypass systemd-resolved and observe that there is nothing wrong with the network:
```
# dig -t a github.com @8.8.8.8 +short
15:12:37.155086 IP 10.x.x.39.39697 > 8.8.8.8.53: 21350+ [1au] A? github.com. (51)
15:12:37.169174 IP 8.8.8.8.53 > 10.x.x.39.39697: 21350 1/0/1 A 140.82.118.3 (55)
140.82.118.3
```
- Check logs:
```
# /etc/systemd/system/systemd-resolved.service.d/override.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```
```
Apr 22 15:13:45 host systemd-resolved[30017]: Got DNS stub UDP query packet for id 5045
Apr 22 15:13:45 host systemd-resolved[30017]: Looking up RR for github.com IN A.
Apr 22 15:13:45 host systemd-resolved[30017]: Transaction 59633 for <github.com IN A> on scope dns on */* now complete with <network-down> from none (unsigned).
Apr 22 15:13:45 host systemd-resolved[30017]: Freeing transaction 59633.
Apr 22 15:13:45 host systemd-resolved[30017]: Sending response packet with id 5045 on interface 1/AF_INET.
Apr 22 15:13:45 host systemd-resolved[30017]: Processing query...
```
Interestingly, `netplan apply` will alter `/run/systemd/resolve/resolv.conf` to the DNS servers found on the `lo` device (*). So resolving does work if you're not using 127.0.0.53.

(*) The third mode of operation:

> systemd-resolved maintains the /run/systemd/resolve/resolv.conf file for compatibility with traditional Linux programs.

However, a chrooted postfix on this box got an older resolv.conf with 127.0.0.53 in it and has not been able to send any mail because of the SERVFAILs.

- Can you explain why it appears that systemd-resolved thinks the network is down?
- Do you need any other info to aid in debugging this?

**Comment by poettering** [comment]
how does "resolvectl status" look like on this host?

**Comment by wdoekes** [comment]
> Use resolvectl status ( systemd-resolve --status when using systemd version earlier than 239) to show your global and per-link DNS settings.

```
# systemd-resolve --status
Global
         DNS Servers: 8.8.8.8
          DNSSEC NTA: 10.in-addr.arpa
                      16.172.in-addr.arpa
                      168.192.in-addr.arpa
                      17.172.in-addr.arpa
                      18.172.in-addr.arpa
                      19.172.in-addr.arpa
                      20.172.in-addr.arpa
                      21.172.in-addr.arpa
                      22.172.in-addr.arpa
                      23.172.in-addr.arpa
                      24.172.in-addr.arpa
                      25.172.in-addr.arpa
                      26.172.in-addr.arpa
                      27.172.in-addr.arpa
                      28.172.in-addr.arpa
                      29.172.in-addr.arpa
                      30.172.in-addr.arpa
                      31.172.in-addr.arpa
                      corp
                      d.f.ip6.arpa
                      home
                      internal
                      intranet
                      lan
                      local
                      private
                      test

Link 6 (docker0)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 5 (enp67s0f1)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 4 (enp67s0f0)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 3 (eno2np1)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 2 (eno1np0)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
```
(And the 8.8.8.8 is in there because of DNS= setting in `/etc/systemd/resolved.conf`. Without it, --status would list no DNS at all.)

netplan config:
```
network:
  version: 2
  renderer: networkd
  ethernets:
    enp67s0f1: {}
    enp67s0f0: {}
    lo:
      nameservers:
        addresses:
        - 8.8.8.8
```
`ip route`:
```
default proto bgp metric 20 
	nexthop via inet6 fe80::xxxx:xxxx:xxxx:f5e0 dev enp67s0f0 weight 1 
	nexthop via inet6 fe80::xxxx:xxxx:xxxx:c1f dev enp67s0f1 weight 1 
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown 
```
`route -n`:
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         0.0.0.0         0.0.0.0         UG    20     0        0 enp67s0f0
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
```
`ip a show enp67s0f0`:
```
4: enp67s0f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 98:xx:xx:xx:xx:1e brd ff:ff:ff:ff:ff:ff
    inet6 fe80::xxxx:xxxx:xxxx:7a1e/64 scope link 
       valid_lft forever preferred_lft forever
```
`# vtysh -c 'show bgp ipv4 unicast summary'`:
```
BGP router identifier 10.x.x.39, local AS number 65xxx vrf-id 0
BGP table version 3
RIB entries 2, using 368 bytes of memory
Peers 2, using 41 KiB of memory
Peer groups 1, using 64 bytes of memory

Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
enp67s0f0       4      65yyy    4996    4996        0    0    0 04:09:39            1
enp67s0f1       4      65yyy    4995    4994        0    0    0 04:09:33            1

Total number of neighbors 2
```

**Comment by poettering** [comment]
so resolved thinks none of the ifaces are configured with a global IP address. What does "ip addr" output for you?

**Comment by wdoekes** [comment]
I already supplied the relevant `ip a(addr)` output. But here's the full one:
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet 10.x.x.39/32 brd 10.x.x.39 scope global lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eno1np0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 00:xx:xx:xx:xx:f2 brd ff:ff:ff:ff:ff:ff
3: eno2np1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 00:xx:xx:xx:xx:f3 brd ff:ff:ff:ff:ff:ff
4: enp67s0f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 98:xx:xx:xx:xx:1e brd ff:ff:ff:ff:ff:ff
    inet6 fe80::xxxx:xxxx:xxxx:7a1e/64 scope link 
       valid_lft forever preferred_lft forever
5: enp67s0f1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 98:xx:xx:xx:xx:1f brd ff:ff:ff:ff:ff:ff
    inet6 fe80::xxxx:xxxx:xxxx:7a1f/64 scope link 
       valid_lft forever preferred_lft forever
6: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 02:xx:xx:xx:xx:ca brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
```

**Comment by poettering** [comment]
in that list there's no outbound interface that has both a global address assigned *and* is up. The only iface with an IP address outside of the link-local range is docker0, but it's not up. and the ifaces that are UP have no global address assigned, or are the loopback device

Hence resolved correctly detects that networking is not available, no?

**Comment by hbos** [comment]
The global address is on the loopback device (`lo`), `inet 10.x.x.39/32` and is distributed by BGP.

It uses BGP unnumbered interfaces where it advertises an IPv4 route with an IPv6 next hop. The BGP sessions are established with the neighbor over the auto-configured ipv6 link local address. So one does not need IPv4 address configuration on links. Which is really nice for automation and simplifies IP address management.

For reference, these [docs](https://docs.cumulusnetworks.com/cumulus-linux-41/Layer-3/Border-Gateway-Protocol-BGP/#bgp-unnumbered-interfaces) explain it a bit better.

**Comment by poettering** [comment]
So what should resolved do? If there's an IP address of global scope on "lo" then assume we are online, too?

**Comment by wdoekes** [comment]
Good question. I don't have the definitive answer.

> If there's an IP address of global scope on "lo" then assume we are online, too?

That would work, and is probably easiest.

You could use the fact that `/proc/net/route` has a 0.0.0.0/0 entry (even though the gateway has no v4 IP):
```
# route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         0.0.0.0         0.0.0.0         UG    20     0        0 enp33s0f0
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0

# awk '{iface=$1;dest=$2;mask=$8; if(dest=="00000000"&&mask=="00000000"){print iface}}'  /proc/net/route 
enp33s0f0
```
Although, that would make everyone offline who don't have a 0.0.0.0/0 route. (LAN-only machines? In fact, perhaps the upstream DNS server is running in a docker container running on localhost :scream: )

You could ask `ip route`:
```
# ip -4 route get 1.2.3.4
1.2.3.4 via inet6 fe80::xxxx:xxxx:xxxx:778 dev enp33s0f1 src 10.x.x.39 uid 0 
    cache 
```
Or do what `ip route` does:
```
socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 3
setsockopt(3, SOL_SOCKET, SO_SNDBUF, [32768], 4) = 0
setsockopt(3, SOL_SOCKET, SO_RCVBUF, [1048576], 4) = 0
setsockopt(3, SOL_NETLINK, NETLINK_EXT_ACK, [1], 4) = 0
bind(3, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = 0
getsockname(3, {sa_family=AF_NETLINK, nl_pid=565, nl_groups=00000000}, [12]) = 0

sendmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 
  msg_namelen=12, msg_iov=[{iov_base={{len=36, type=RTM_GETROUTE, 
  flags=NLM_F_REQUEST, seq=1588061342, pid=0}, {rtm_family=AF_INET, rtm_dst_len=32, 
  rtm_src_len=0, rtm_tos=0, rtm_table=RT_TABLE_UNSPEC, rtm_protocol=RTPROT_UNSPEC, 
  rtm_scope=RT_SCOPE_UNIVERSE, rtm_type=RTN_UNSPEC, 
  rtm_flags=RTM_F_LOOKUP_TABLE}, {{nla_len=8, nla_type=RTA_DST}, 1.2.3.4}}, iov_len=36}], 
  msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 36

recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 
  msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, 
  msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 128

recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 
  msg_namelen=12, msg_iov=[{iov_base={{len=128, type=RTM_NEWROUTE, flags=0, 
  seq=1588061342, pid=565}, {rtm_family=AF_INET, rtm_dst_len=32, rtm_src_len=0, rtm_tos=0, 
  rtm_table=RT_TABLE_MAIN, rtm_protocol=RTPROT_UNSPEC, 
  rtm_scope=RT_SCOPE_UNIVERSE, rtm_type=RTN_UNICAST, rtm_flags=RTM_F_CLONED}, 
  [{{nla_len=8, nla_type=RTA_TABLE}, RT_TABLE_MAIN}, {{nla_len=8, nla_type=RTA_DST}, 
  1.2.3.4}, {{nla_len=8, nla_type=RTA_OIF}, if_nametoindex("enp33s0f1")},
  {{nla_len=8, nla_type=RTA_PREFSRC}, 10.x.x.39}, {{nla_len=22, nla_type=RTA_VIA}, 
  {rtvia_family=AF_INET6, inet_pton(AF_INET6, "fe80::xxxx:xxxx:xxxx:778", &rtvia_addr)}}, 
  {{nla_len=8, nla_type=RTA_UID}, 0}, {{nla_len=36, nla_type=RTA_CACHEINFO}, 
  {rta_clntref=5, rta_lastuse=33734946, rta_expires=0, rta_error=0, rta_used=0, rta_id=0, 
  rta_ts=0, rta_tsage=0}}]}, iov_len=128}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 128
```
But for that `ip route` query to succeed you'd need an IP that you know _isn't_ yours... so that's probably not the best of ideas... (Unless of course you use the upstream DNS IP, which is what you're after anyway.)

I'm sorry that I don't have an easy answer here.

**Comment by poettering** [comment]
Should be fixed since 90bdc8be66765df09bbc355783cee7204a5ebb31. Closing.
